include/mbedtls/mps/writer.h

```diff
@@ -1,50 +1,63 @@
- * After that, the provider feeds another outgoing data buffer
- * to the writer and the cycle starts again.
- * In the event that a consumer's request cannot be fulfilled
- * on the basis of the outgoing data buffer provided by the
- * provider, the writer may provide a temporary 'queue' buffer
- * instead. In this case, the queue buffer will be copied to the
- * outgoing data buffers when the provider subsequently provides
- * them. The details of this are opaque to the consumer and the
- * provider, but it means that if the provider feeds an outgoing
- * data buffer to the writer, the writer might entirely fill it
- * immediately on the basis of what has been queued internally.
- *
- * From the perspective of the consumer, the state of the
- * writer is a potentially empty list of output buffers that
- * the writer has provided to the consumer.
- * New buffers can be requested through calls to mbedtls_writer_get(),
- * while previously obtained output buffers can be marked processed
- * through calls to mbedtls_writer_commit(), emptying the list of
- * output buffers and invalidating them from the consumer's perspective.
+ * Writers provide a means of communication between
+ * - a 'provider' supplying buffers to hold outgoing data, and
+ * - a 'consumer' writing data into these buffers.
+ * Both the size of the data buffers the provider prepares and the size of
+ * chunks in which the consumer writes the data are variable and may be
+ * different. It is the writer's responsibility to do the necessary copying
+ * and pointer arithmetic.
+ *
+ * For example, the provider might be the [D]TLS record layer, offering
+ * to protect and transport data in records of varying size (depending
+ * on the current configuration and the amount of data left in the current
+ * datagram, for example), while the consumer would be the handshake logic
+ * layer which needs to write handshake messages. The size of handshake
+ * messages are entirely independent of the size of records used to transport
+ * them, and the writer helps to both split large handshake messages across
+ * multiple records, and to pack multiple small handshake messages into
+ * a single record. This example will be elaborated upon in the next paragraph.
+ *
+ * Basic flow of operation:
+ * First, the provider feeds an outgoing data buffer to the writer, transferring
+ * it from 'providing' to 'consuming' mode; in the example, that would be record
+ * layer providing the plaintext buffer for the next outgoing record. The
+ * consumer subsequently fetches parts of the buffer and writes data to them,
+ * which might happen multiple times; in the example, the handshake logic
+ * layer might request and fill a buffer for each handshake message in the
+ * current outgoing flight, and these requests would be served from successive
+ * chunks in the same record plaintext buffer if size permits. Once the consumer
+ * is done, the provider revokes the writer's access to the data buffer,
```

Comment:
I think this sentence lacks a connection between the consumer being done and the provider revoking the access to the buffer. I believe that is the commit operation. I'd explain it in two stages, e.g.: Once the consumer is done, it signals it to the writer. The provider may then reclaim the data that has been committed to the writer.